### PR TITLE
Issue #2254 - Add Test Mode Per Form Settings

### LINKED ISSUE #2254: Enable Test Mode Per Form
## Issue Overview
Currently, we can enable `Test Mode` globally. However, We can have such feature that a site admin can enable it per form.

I have discussed this point on last Weekly Developer Call and decided to create an issue for further discussion.

**Here are few questions:**
1. Do we have any existing snippet for managing Test Mode Per Form?
2. What do you think this could be a Core feature, Snippet or an Addon?
3. Do we got any Support Requests for this feature till date?
4. What do you think that this feature can be useful?

Let me know your thoughts.

### PR DESCRIPTION
**Pull request by mehul0810**
## Description
This PR resolves #2254 

## How Has This Been Tested?
I have tested this manually with all the 3 options:
1. Global
2. Enable Test Form Per Form
3. Disable Test Form Per Form

## Screenshots (jpeg or gifs if applicable):
image

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by mehul0810** [comment]
@ravinderk @DevinWalker I've add update icon as an icon for Test Mode tab. Let me know if i need to change that icon.

File: includes/misc-functions.php [hunk 1]
```diff
@@ -23,9 +23,20 @@
  */
 function give_is_test_mode() {
```
Comment (ravinderk) [inline includes/misc-functions.php]: @mehul0810 As discussed over call please improve this function and check multiple scenarios like:
1. Donation form page
2. Donation form shortcode on page
3. Multiple donation form shortcode on page

**Comment by mehul0810** [comment]
@ravinderk I've completed this PR with the required changes including unit tests.

Also, I've taken care of the case where there are multiple shortcodes added to a page or post or custom post type.

**Here are the areas, I have tested:**
1. Single Donation Page with per form settings
2. Single Donation Page with global settings
3. Shortcodes with per form settings
4. Shortcodes with global settings
5. Multiple shortcodes on a single page.

**Please check the video for more details:** https://drive.google.com/open?id=1DHeguCwGLCXfrAIDTqB4lLpOCjUXbtg4

File: includes/forms/template.php [hunk 2]
```diff
@@ -1802,16 +1802,22 @@ function give_filter_success_page_content( $content ) {
...
-function give_test_mode_frontend_warning() {
+function give_test_mode_frontend_warning( $form_id ) {
 
-	if ( give_is_test_mode() ) {
-		echo '<div class="give_error give_warning" id="give_error_test_mode"><p><strong>' . esc_html__( 'Notice:', 'give' ) . '</strong> ' . esc_html__( 'Test mode is enabled. While in test mode no live donations are processed.', 'give' ) . '</p></div>';
+	if ( give_is_test_mode( $form_id ) ) {
+		echo sprintf(
+			'<div class="give_error give_warning" id="give_error_test_mode"><p><strong>%1$s</strong>%2$s</p></div>',
+			__( 'Notice:', 'give' ),
+			__( 'Test mode is enabled. While in test mode no live donations are processed.', 'give' )
```
Comment (DevinWalker) [inline includes/forms/template.php]: Space needed here between "Notice:" and "Test mode..."

**Comment by DevinWalker** [comment]
@mehul0810 I just watched your video and the functionality looks nice but I do have a few concerns.

1. When the "Test" gateway is not active globally but "Test Mode" is enabled on a form we have to use actual gateways to test. Sometimes these gateways may not be enabled properly for testing (missing API keys). It seems to me that if an we enable the "Test" gateway for a form we also need to enable the "Test" gateway.

2.  I'm concerned that the per-form "Test Mode" tab isn't relevant for the majority of our users and will just add unnecessary weight to the UI and may confuse some admins. What I would like to see is an option under the "Advanced" tab within global Settings that allow admins to enable or disable "Per Form Test Mode". 

Description:
> When enabled a new "Test Mode" tab will display within each donation form's settings that allows you to toggle test mode for just that form.

**Comment by DevinWalker** [comment]
@mehul0810 thank you for your hard work at this but after reviewing and discussing with the team it's just not a functionality that would be very useful to our users and for support as well. Additionally it would require some additional updates to our gateways to pass `give_is_test_mode()` with the new $form_id param. I'm going to close this one out as well as the issue but I appreciate the work.

**Pull request closed without merging**